### PR TITLE
feat: add custom key for cel expression support

### PR DESCRIPTION
**Pull request by TzlilSwimmer123**
Screen Shot 2023-07-19 at 15 09 31

### CODE REVIEW

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	return nil, nil
+}
+
+func (customKeyCELDefinitionSchema CustomKeyCELDefinitionSchema) Validate(ctx jsonschema.ValidationContext, dataValue interface{}) error {
+	CELDefinitionSchema, err := convertCustomKeyCELDefinitionSchemaToCELDefinitionSchema(customKeyCELDefinitionSchema)
+	if err != nil {
+		return ctx.Error(CustomKeyValidationErrorKeyPath, err.Error())
+	}
+	// wrap dataValue (the resource that should be validated) inside a struct with parent object key
+	dataValueObjectWrapper := make(map[string]interface{})
```
Comment (shmu3l) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: naming maybe `resourceWithParentKey` ?

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			return nil, fmt.Errorf("CELExpression must be an object of type CELExpression %s", err.Error())
+		}
+		CELDefinition.CELExpressions = append(CELDefinition.CELExpressions, CELExpression)
+	}
+
+	return &CELDefinition, nil
+}
+
+func getCELEnvInputs(dataValue map[string]interface{}) ([]cel.EnvOption, error) {
+	var inputs map[string]any
```
Comment (shmu3l) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: `var inputs map[string]interface{}`

**Review by shmu3l — DISMISSED** [review]

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 2 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: I think that it suppose to be any because it can be any type

Comment (royhadad) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: `any` is an alias for `interface{}` :)

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
+// This file defines a custom key to implement the logic for rego rule:
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]:
```suggestion
// This file defines a custom key to implement the logic for a CEL rule:
```

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 2 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: Good to know:)

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			"type": "array"
+		}
+	}
+}`)
+
+func (CustomKeyCELDefinitionCompiler) Compile(ctx jsonschema.CompilerContext, m map[string]interface{}) (jsonschema.ExtSchema, error) {
+	if customKeyCELRule, ok := m[CELDefinitionCustomKey]; ok {
+		customKeyCELRuleObj, validObject := customKeyCELRule.([]interface{})
+		if !validObject {
+			return nil, fmt.Errorf("CELDefinition must be an object")
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]:
```suggestion
			return nil, fmt.Errorf("CELDefinition must be an array")
```

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		if res1.Type().TypeName() != "bool" {
+			return ctx.Error(CustomKeyValidationErrorKeyPath, "cel expression needs to return a boolean")
+		}
+
+		celReturnValue, ok := res1.Value().(bool)
+		if !ok {
+			return ctx.Error(CustomKeyValidationErrorKeyPath, "cel expression needs to return a boolean")
+		}
+		if !celReturnValue {
+			return ctx.Error(CustomKeyValidationErrorKeyPath, "cel expression failure message: %s", celExpression.Message)
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: "cel expression failure message:"
does @adifayer want this prefix for every CEL failure? how do users experience it?
maybe include a screenshot

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+type CELDefinition struct {
+	CELExpressions []CELExpression
+}
+
+func convertCustomKeyCELDefinitionSchemaToCELDefinitionSchema(CELDefinitionSchema CustomKeyCELDefinitionSchema) (*CELDefinition, error) {
+	var CELDefinition CELDefinition
+	for _, CELExpressionFromSchema := range CELDefinitionSchema {
+		var CELExpression CELExpression
+		b, err := json.Marshal(CELExpressionFromSchema)
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go:112]: why `Marshal` and then `Unmarshal`? if you just need a type conversion then use Go's type conversion. Am I missing something?

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+		CELDefinition.CELExpressions = append(CELDefinition.CELExpressions, CELExpression)
+	}
+
+	return &CELDefinition, nil
+}
+
+func getCELEnvInputs(dataValue map[string]interface{}) ([]cel.EnvOption, error) {
+	var inputs map[string]any
+	dataValueBytes, _ := json.Marshal(dataValue)
+	if err := yaml.Unmarshal(dataValueBytes, &inputs); err != nil {
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: Let's see if I understand this properly.
- CEL expects the input to be a yaml
- you get it as a Go object
- therefore, you need to convert the object to a yaml
is that correct?

**Review by royhadad — DISMISSED** [review]

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 6 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go:114]: Unfortunately Go's type conversion is not helpful here.  I will try just to make sure I am not mistaking

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 7 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: Your comment made me understand it was redundant! thanks

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 6 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go:112]: Ok let's see, what is your suggestion?

File: pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go (hunk 5 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyCELDefinition.go]: I will show it to Adi

**Review by hadar-co — APPROVED** [review]

**Pull request merged**